### ISSUE TITLE
Saving and loading can sometimes change a Logic block's outcome.

### ISSUE THREAD

**Issue by NoExit4U** [Anuken/Mindustry#4888]
**Platform**: Windows

**Build**: Steam build 126.1

**Issue**: When saving a game and then loading it, the outcome of a logic block can be different. In this case it causes disabled Factories to sometimes become permanently active after loading the game.

Here you see that the 3 factories are disabled, even when I wait 5+ minutes. The Factory in the top left is crossed because it never showed the bug even though it have very similar code.
![20210307165205_1](https://user-images.githubusercontent.com/75215867/110248262-7ae7ac00-7f70-11eb-8081-dbc5904166c3.jpg)

After saving and loading twice, one of the factories immediately became active
![20210307165238_1](https://user-images.githubusercontent.com/75215867/110248265-7d4a0600-7f70-11eb-8886-4e4efc732fe6.jpg)

After Saving and Loading about 10 times, I managed to get all 3 of the factories to become active.
![20210307165359_1](https://user-images.githubusercontent.com/75215867/110248268-7f13c980-7f70-11eb-8d07-118a24fdbe3e.jpg)

**Steps to reproduce**:
- Play the map in Sandbox or Attack (both were tested). See below for the map file.
- Stand on the spot as seen on the screenshot, near (197, 169).
- Observe that the 4 factories are not producing units even though they have enough resources.
- (Optionally) Let any amount of time pass and observe that they won't activate on their own.
- Save and Quit.
- Load the saved file again (I did it without closing Mindustry).
- If one or more Unit Factories have started producing then the bug has occured. If not repeat Save and Quit, then load until it occurs.

It may require a couple of tries before the bug works. This is my record of how many factories started to produce while testing multiple times:
2 - 2 - 0 - 0 - 0 - 0 - 1 - 1 - 1 - 1 - 0 - 1 - 0 - 0 - 0 - 1 - 0 - 0 - 1 - 1 - 0 - 1

**Link(s) to mod(s) used**: None (the save file may have been with the Cliff mod active, but I was able to reproduce the bug even when I disabled it)

**Additional Info**
The logic scripts are in a chain as follows:
- Logic Block 1 => Memory Cell => Logic Block 2 (You can spot them because they have nearby Surge Walls)

Logic Block1 writes 1 into the Memory Cell.
Logic Block1 detects whether or not their linked blocks died. If so, it writes 0 into the Memory Cell.

Logic Block2 loops to wait until Memory Cell becomes 1
Logic Block2 loops to wait until Memory Cell becomes 0 again (while disabling the Factory).
Logic Block2 then goes into a loop where it enables the factory forever.

This setup is not guaranteed for it to go bad. Out of the 4 setups, I have only seen 3 of them show the bug.
The logic code is not identical for all Logic blocks. For example for the 2 factories on the right, their "Logic Block2" will also check if nearby blocks died to activate.

Other logic blocks on the map appear to have similar issues but are harder to observe and have more complex logic blocks.

**Other observations**
- I never seen a Factory that was active go back to become inactive. Either by waiting or by loading the same file again.

**Map File**: [map_91.zip](https://github.com/Anuken/Mindustry/files/6097597/map_91.zip)
**Save file**: A Save for when all 3 factories have been activated: [411.zip](https://github.com/Anuken/Mindustry/files/6097593/411.zip)

---

*Place an X (no spaces) between the brackets to confirm that you have read the line below.*  
- [X] **I have updated to the latest release (https://github.com/Anuken/Mindustry/releases) to make sure my issue has not been fixed.**
- [X] **I have searched the closed and open issues to make sure that this problem has not already been reported.**

**Comment by Anuken** [comment]
It sounds like you're relying on update order for your logic blocks; when the map reloads, update order will change unpredictably. This is not a bug. Logic should be designed with this in mind.

**Comment by NoExit4U** [comment]
It can't be an update issue:

1) The loop in Logic Block 2 waits for Memory Cell 1 to become 1 first.

2) If it was an "update order" issue then I would have seen one of these factories become active when I start the map, which didn't happen in more than 25 times. Only after saving and loading does it happen.

3) I changed the code of the lower left logic block to:
![20210307192414_1](https://user-images.githubusercontent.com/75215867/110250332-0cf4b200-7f7b-11eb-8ee5-d420a7c318f8.jpg)
and
(cellIndex = 0 didn't fit on the screenshot)
![20210307192434_1](https://user-images.githubusercontent.com/75215867/110250337-11b96600-7f7b-11eb-8ee2-390c6bc22a05.jpg)

Now the order no longer matters and I still managed to replicate the bug after loading a couple of times.

**Comment by Anuken** [comment]
There are so many variables and operations here that it's impossible to tell whether it's a bug in your code or the game. Historically, every "bug" related to large logic programs has turned out to be an issue with the program, not Mindustry. Unless you can produce a minimal example that clearly demonstrates incorrect behavior (and specified *what exactly goes wrong*), I won't be looking into this.

**Comment by NoExit4U** [comment]
Ok, I'll try and isolate the bug. I'll open a new issue if I find it.

Before I go can I know which of these are supposed to be saved in a map?
- Which instruction is currently running in a logic block
- The content of Logic variables
- The content of memory cells

**Comment by Anuken** [comment]
All of those are saved.
